src/package1/CustomArrayList.java

```diff
@@ -124,31 +124,52 @@ public boolean addAll(int index, Collection<? extends String> c) {
     public boolean removeAll(Collection<?> c) {
         int arrayIndex = 0;
 
-        for (String str: this){
-            if(!c.contains(str)){
+        for (String str : this) {
+            if (!c.contains(str)) {
                 elements[arrayIndex] = str;
                 arrayIndex++;
             }
-         }
+        }
         size = arrayIndex;
         return true;
     }
 
     @Override
     public boolean retainAll(Collection<?> c) {
         int arrayIndex = 0;
-        String [] a = new String[elements.length];
-
-        for (String sElements : elements) {
-
-            for (Object sCollection : c) {
-                if (sElements.equals(sCollection)){
-                    a[arrayIndex] = sElements;
-                    arrayIndex++;
-                }
+        for (int i = 0; i < size; i++) {
+            // Checking of containing of elements
+            // and getting a start index of appropriation.
+            if (c.contains(elements[i])) {
+                arrayIndex = i;
+                break;
+            }
+            // Checking of ending and exit of method.
+            // It means that collection hasn't coincidences.
+            // Create an empty array.
+            if (i == size - 1) {
+                clear();
+                return true;
             }
         }
-        elements = a;
+        // We assign a first element to thisArray if the method comes here.
+        //Create a new counter of a new thisArray.
+        elements[0] = elements[arrayIndex];
+        int newArrayIndex = 1;
+
+
+        // We're going into a remainder of previous thisArray by using a counter "i".
+        // We assign the elements are contained in a collection for a new thisArray.
+        for (int i = arrayIndex + 1; i < size; i++) {
+            if (c.contains(elements[i])) {
+                elements[newArrayIndex] = elements[i];
+                newArrayIndex++;
```

Comment:
newArrayIndex is equals or less of "i" and elements of array copy into equals or less position always.